### ISSUE TITLE
dots in usernames verboten

### ISSUE THREAD

**Issue by jstange** [chef/chef-server#557]
I'm experimenting with LDAP functionality, bolting our standalone Chef server installation (12.2.0, plus some older versions I no longer recall) onto an existing Active Directory domain. The weird restrictions on usernames are getting in my way, specifically the ban on dots, like `john.stange`. That's a fairly standard corporate convention for usernames. What's the rationale?

Just so we're clear what I'm talking about, asking the REST API to create a user by POSTing the following to `/users` results in a 400 (Bad Request):

    {
     "username": "john.stange",
     "first_name": "John",
     "middle_name": "LDAP",
     "last_name": "Stange",
     "display_name": "John Stange LDAP",
     "email": "john.stange@eglobaltech.com",
     "password": "sometestnonsense"
    }

Strip out the dots (set username to `johnstange`) and it happily complies.

**Comment by stevendanna** [comment]
I would need to do some archeology on the origin of that restriction.  However, the chef username and LDAP user names don't need to be exactly the same.  If you use Chef Manage to link an LDAP account to a chef account, it typically replaces invalid characters with '_'.  However, the user is still able to login using their unmodified LDAP username.

**Comment by marcparadise** [comment]
I'd like to track that down  - we've had that restriction as far back as I can remember, but I don't know that we ever captured why. I do know that we don't do any parsing of username in a meaningful way within erchef, so [pending researching and checking to make sure that there's nothing on the management console side of things] I'm inclined to remove the restriction

**Comment by jstange** [comment]
Eeeeenteresting. Any pointers on how to get at that LDAP linking from the API? I'm not seeing it in my skim of https://docs.chef.io/api_chef_server.html, but I'm probably just looking for the wrong verb.

I've got a nickel on "didn't feel like accounting for weird characters in REST paths."

**Comment by stevendanna** [comment]
> Any pointers on how to get at that LDAP linking from the API?

If you update the user record or create the user record, you can set `external_authentication_uid`.  The value of external_authentication_uid should map to the value of the login_attribute in your LDAP directory.  By default the login_attribute is sAMAccountName.  That is, if your login_attribute is configured to be 'uid' and user John has a uid of "john.johnson", you can create a chef user called john_johnson with the external_authentication_uid of "john.johnson". That user would be able to log into the management console with the "username" of john.johnson and their LDAP password.

Note, that if you are trying to update an existing user record to add an external_authentication_uid, you have to be the superuser or "pivotal" user.  They key for "pivotal" can be found in /etc/opscode on your chef server.

Given your interested in the API, I feel compelled to point out that LDAP really *only* affects Management Console logins.  All API requests still happen via the public-private keypairs and don't contact LDAP.

**Comment by jstange** [comment]
Yeah, we worked that out in the midst of yesterday's fiddling. We're configuring a bunch of stuff bundled with our management suite for AD compatibility, including Chef. The latter's a placeholder exercise at this point, since we currently don't even use the UI.

**Comment by marcparadise** [comment]
Related: #676

**Comment by docwhat** [comment]
I ran across this by trying to use `mail` as the `loginAttribute` (e.g. the email address).

What happened was:

* I logged in the server with `docwhat@us.example.com`.
* The chef username was `docwhat_us_example_com` (that's fine).
* However, when using supermarket, I couldn't login via either `docwhat@us.example.com` nor `docwhat_us_example_com`.

I suspect that there is something weird in the way the login for the `oc-id` endpoint works that caused the mis-matched email, LDAP `loginAttribute`, and the chef-server username confusion for the poor thing.
